### PR TITLE
add leaderboard caching and pagination

### PR DESCRIPTION
**Pull request by ginkoid**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
 const path = require('path')
 const express = require('express')
 
+require('./server/leaderboard')
```
Comment (chen-robert) [inline app.js]: Is require with side-effects good practice? Maybe add a function to call like `.start()`

File: server/cache/leaderboard.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+const getRange = async ({ start, end }) => {
+  let redisResult = []
+  try {
+    redisResult = await redisLrange('leaderboard', start * 3, end * 3 - 1)
+  } catch (e) {}
+  const result = []
+  for (let i = 0; i < redisResult.length / 3; i++) {
+    result.push([redisResult[i], redisResult[i + 1], parseInt(redisResult[i + 2])])
```
Comment (chen-robert) [inline server/cache/leaderboard.js]: The meaning of `redisValue[i + {1,2}]` should be documented somewhere? Maybe in a comment or a javascript class

File: server/leaderboard/calculate.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  const solvedChalls = userSolves.get(userId)
+  for (let j = 0; j < solvedChalls.length; j++) {
+    // Add the score for the specific solve loaded fr om the challengeValues array using ids
+    currScore += challengeValues.get(solvedChalls[j])
+  }
+  userScores.push([userId, users[i].name, currScore, userLastSolves.get(userId)])
+}
+
+const sortedUsers = userScores.sort((a, b) => {
+  const scoreCompare = b[2] - a[2]
```
Comment (chen-robert) [inline server/leaderboard/calculate.js]: Is there a way to make `b[2]` more informative? It's not really clear what property this is referring to.

Comment (jimmyl02) [reply server/leaderboard/calculate.js]: a and b refer to elements in userScores but we could document the type of data stored in the array

ex. 
```
/**
 * @typedef {Object[]} userScores - a list of users and their scores
 * @typedef {string} userScores[0] -  userId
 * @typedef {string} userScores[1] -  username
 * @typedef {string} userScores[2] -  last solve time
 */
```

**Pull request merged**